pkg/virt-handler/vm.go

```diff
@@ -353,8 +341,18 @@ func (d *VirtualMachineController) execute(key string) error {
 	if err != nil {
 		return err
 	} else if watchdogExpired && vm.IsRunning() {
-		log.Log.Object(vm).V(3).Info("Shutting down due to expired watchdog.")
+		log.Log.Object(vm).Info("Shutting down due to expired watchdog.")
 		shouldShutdownAndDelete = true
+		if domainExists {
+			// Virt-launcher provids domain state. If virt-launcher
+			// is down, the domain needs to be deleted from the cache.
+			err = d.domainInformer.GetStore().Delete(domain)
+			if err != nil {
+				return err
```

Comment:
Under normal operation, the listwatcher will process all of this correctly. Delete events come in and domains are removed from the cache by the informer. 

This is an edge case where the informer never receives the Delete event from the VM Pod. The watchdog expires and we remove the cache entry here.  This condition could occur if virt-handler is down when the final delete notification is sent, or if virt-launcher forcibly exits in a way that results in the final delete notification not being sent.

Reply:
> This condition could occur if virt-handler is down when the final delete notification is sent or if virt-launcher forcibly exits in a way that results in the final delete notification not being sent.

That should normally not be necessary. A few mechanisms should be sufficient to prevent that:

 * warming up the domain cache with the vm cluster-state. If we then don't find a domain, matching the expectation based on the vm, the informer should delete the object from the cache automatically
 * If we know which socket belongs to which vm, we can infer from that, that if the socket connection gets closed, or if no one listens on the socket, which vm got removed.

@davidvossel can you think of a scenario where we can miss a delete if the above conditions are met?

Reply:
i'll give this part some more attention this afternoon. There were some things i was trying to avoid with regards to coupling the client connection to the domain's status. I think this can be simplified though.

Reply:
@rmohr 

I've removed the watchdog informer entirely. The domain informer now polls for stale watchdog files and fires a delete event when one is encountered.

I'm hesitant to tie any of this Delete event logic directly to the unix socket, which is why i'm still using the watchdog files. I don't want the presence, or lack of presence of the unix socket file to infer a stale domain needs to be processed.

Reply:
works for me. Just modifying the cache outside of the informer is a no go afaik (locking, ...).